src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,858 @@
+   * TableStatistics
+   */
+  std::vector<std::shared_ptr<BaseColumnStatistics>> column_statistics{
+      left_input_table_statistics.column_statistics.size() + right_input_table_statistics.column_statistics.size()};
+
+  const auto left_column_count = left_input_table_statistics.column_statistics.size();
+  for (auto column_id = ColumnID{0}; column_id < left_column_count; ++column_id) {
+    column_statistics[column_id] = left_input_table_statistics.column_statistics[column_id]->scaled(left_selectivity);
+  }
+
+  for (auto column_id = ColumnID{0}; column_id < right_input_table_statistics.column_statistics.size(); ++column_id) {
+    column_statistics[left_column_count + column_id] =
+        right_input_table_statistics.column_statistics[column_id]->scaled(right_selectivity);
+  }
+
+  const auto row_count = Cardinality{left_selectivity * right_selectivity};
+
+  return std::make_shared<TableStatistics>(std::move(column_statistics), row_count);
+}
+
+template <typename T>
+std::shared_ptr<GenericHistogram<T>> CardinalityEstimator::estimate_inner_equi_join_with_histograms(
+    const AbstractHistogram<T>& left_histogram, const AbstractHistogram<T>& right_histogram) {
+  /**
+   * left_histogram and right_histogram are turned into "unified" histograms by `split_at_bin_bounds`, meaning that
+   * their bins are split so that their bin boundaries match.
+   * E.g., if left_histogram has a single bin [1, 10] and right histogram has a single bin [5, 20] then
+   * unified_left_histogram == {[1, 4], [5, 10]}
+   * unified_right_histogram == {[5, 10], [11, 20]}
+   * The estimation is performed on overlapping bins only, e.g., only the two bins [5, 10] will produce matches.
+   */
+
+  auto unified_left_histogram = left_histogram.split_at_bin_bounds(right_histogram.bin_bounds());
+  auto unified_right_histogram = right_histogram.split_at_bin_bounds(left_histogram.bin_bounds());
+
+  auto left_idx = BinID{0};
+  auto right_idx = BinID{0};
+  auto left_bin_count = unified_left_histogram->bin_count();
+  auto right_bin_count = unified_right_histogram->bin_count();
+
+  GenericHistogramBuilder<T> builder;
+
+  // Iterate over both unified histograms and find overlapping bins
+  for (; left_idx < left_bin_count && right_idx < right_bin_count;) {
+    const auto left_min = unified_left_histogram->bin_minimum(left_idx);
+    const auto right_min = unified_right_histogram->bin_minimum(right_idx);
+
+    if (left_min < right_min) {
+      ++left_idx;
+      continue;
+    }
+
+    if (right_min < left_min) {
+      ++right_idx;
+      continue;
+    }
+
+    DebugAssert(unified_left_histogram->bin_maximum(left_idx) == unified_right_histogram->bin_maximum(right_idx),
+                "Histogram bin boundaries do not match");
+
+    // Overlapping bins found, estimate the join for these bins' range
+    const auto [height, distinct_count] = estimate_inner_equi_join_of_bins(  // NOLINT
```

Comment:
This looks a lot like column-vs-column. Deduplicate and pass "bin function thingy" as a template parameter?

Reply:
Nope, imo not worth it at all. You wanna be able to play around with what column-vs-column does without worrying to break inner equi join estimation. At least I want to.